plugin/lspconfig.lua

```diff
@@ -0,0 +1,84 @@
+local version_info = vim.version()
+if version_info.minor < 7 then
+  local warning_str = string.format(
+    '[lspconfig] requires neovim 0.7 or later. Detected neovim version: 0.%s.%s',
+    version_info.minor,
+    version_info.patch
+  )
+  vim.notify_once(warning_str)
+  return
+end
+
+local util = require 'lspconfig.util'
+local configs = require 'lspconfig.configs'
+
+local lsp_get_active_client_ids = function()
+  return vim.tbl_map(function(client)
+    return ('%d (%s)'):format(client.id, client.name)
+  end, require('lspconfig.util').get_managed_clients())
+end
+
+local get_clients_from_cmd_args = function(arg)
+  local result = {}
+  for id in (arg or ''):gmatch '(%d+)' do
+    result[id] = vim.lsp.get_client_by_id(tonumber(id))
+  end
+  if vim.tbl_isempty(result) then
+    return util.get_managed_clients()
+  end
+  return vim.tbl_values(result)
+end
+
+-- Called from plugin/lspconfig.vim because it requires knowing that the last
+-- script in scriptnames to be executed is lspconfig.
+vim.api.nvim_create_user_command('LspInfo', function()
+  require 'lspconfig.ui.lspinfo'()
+end, {
+  desc = 'Displays attached, active, and configured language servers',
+})
+
+vim.api.nvim_create_user_command('LspStart', function(info)
+  local server_name = info.fargs[1]
+  if server_name then
+    if configs[server_name] then
+      configs[server_name].launch()
+    end
+  else
+    local buffer_filetype = vim.bo.filetype
+    for _, config in pairs(configs) do
+      for _, filetype_match in ipairs(config.filetypes or {}) do
+        if buffer_filetype == filetype_match then
+          config.launch()
+        end
+      end
+    end
+  end
```

Comment:
```suggestion
  local server_name = info.args and info.args ~= ""
  if server_name then
    if configs[server_name] then
      configs[server_name].launch()
    end
  else
    local other_matching_configs = require('lspconfig.util').get_other_matching_providers(vim.bo.filetype)
    for _, config in ipairs(other_matching_configs) do
      config.launch()
    end
  end
```

there seems to be a bug with using `fargs` and `nargs='?'`

Reply:
Doesn't that suggestion turn ´server_name` into a boolean (in the truthy case, where it matters)?

Reply:
> Doesn't that suggestion turn ´server_name` into a boolean (in the truthy case, where it matters)?

it's specifically that `<f-args>` will get a string when using `nargs=?`, so `fargs[1]` doesn't really do anything useful, since calling `LspStart` without any arguments will give `server_name=""`

Reply:
I meant won't the following expression

```lua
local server_name = info.args and info.args ~= ""
```

always evaluate to a boolean? It's later used to index the `configs` table: `configs[server_name]`.

I'd probably do something like the following:

```lua
if info.args ~= "" then
  local server_name = info.args
  if configs[server_name] then
    configs[server_name].launch()
  end
else
  -- ...
end
```

Reply:
dang 🤦 

I meant to write it like this

```lua
local server_name = info.args ~= "" and info.args
```

but yeah, just need to make sure we're checking that args isn't an empty string